presto-main/src/main/java/com/facebook/presto/metadata/ParametricFunctionBuilder.java

```diff
@@ -107,39 +115,59 @@ public ParametricFunctionBuilder methods(String... methodNames)
     public ParametricFunctionBuilder methods(List<String> methodNames)
     {
         requireNonNull(methodNames, "methodNames is null");
+        methodNames.forEach(this::method);
+        return this;
+    }
 
-        List<Method> matchingMethods = asList(clazz.getMethods()).stream()
-                .filter(method -> methodNames.contains(method.getName()))
-                .collect(toList());
-        List<String> matchingMethodNames = matchingMethods.stream()
-                .map(Method::getName)
+    public ParametricFunctionBuilder method(String methodName)
+    {
+        initMethodGroupIfNeeded();
+        checkState(clazz.isPresent(), "class instance not set");
+        List<NamedMethodHandle> matchingMethods = asList(clazz.get().getMethods()).stream()
+                .filter(method -> methodName.equals(method.getName()))
+                .map(Reflection::methodHandle)
+                .map(methodHandle -> new NamedMethodHandle(methodName, methodHandle))
                 .collect(toList());
+        checkState(!matchingMethods.isEmpty(), "method %s was not found in %s", methodName, clazz);
+        matchingMethods.forEach(currentMethodGroup.get()::addMethod);
+        return this;
+    }
 
-        for (String methodName : methodNames) {
-            checkState(matchingMethodNames.contains(methodName), "method %s was not found in %s", methodName, clazz);
+    private void initMethodGroupIfNeeded()
+    {
+        if (!currentMethodGroup.isPresent() || methodsFlushedForCurrentGroup) {
+            MethodsGroup newMethodGroup = new MethodsGroup();
+            currentMethodGroup = Optional.of(newMethodGroup);
+            methodsGroups.add(newMethodGroup);
+            methodsFlushedForCurrentGroup = false;
         }
+    }
 
-        methodsGroups.add(new MethodsGroup(matchingMethods));
+    public ParametricFunctionBuilder method(String methodName, MethodHandle methodHandle)
+    {
+        initMethodGroupIfNeeded();
+        currentMethodGroup.get().addMethod(new NamedMethodHandle(methodName, methodHandle));
         return this;
     }
 
     public ParametricFunctionBuilder predicate(Predicate<SpecializeContext> predicate)
     {
         requireNonNull(predicate, "predicate is null");
-        checkState(!methodsGroups.isEmpty(), "no methods are selected (call methods() first)");
-        checkState(!getLast(methodsGroups).getPredicate().isPresent(), "predicate already defined for selected methods");
-
-        getLast(methodsGroups).setPredicate(predicate);
+        checkState(currentMethodGroup.isPresent(), "no methods are selected (call methods() first)");
+        checkState(!currentMethodGroup.get().getPredicate().isPresent(), "predicate already defined for selected methods");
+        currentMethodGroup.get().setPredicate(predicate);
+        methodsFlushedForCurrentGroup = true;
```

Comment:
rename methodsFlushedForCurrentGroup to something like methodsSelectedForCurrentGroup